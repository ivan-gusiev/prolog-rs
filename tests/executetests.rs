extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod executetests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::{
        compile::{compile_program, compile_query},
        lang::parse_term,
        run_code,
        util::{case, writeout},
        Machine,
    };

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_execute(input: &str) {
        let query = parse_term(input).unwrap();
        let instructions = compile_query(query);
        let mut machine = Machine::new();
        machine.set_code(&instructions);
        run_code(&mut machine);

        let output = writeout(&machine.iter_heap().collect::<Vec<_>>());
        assert_display_snapshot!(case(
            input,
            output
        ));
    }

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "p(f(X), h(Y, f(a)), Y)",
        "a"
    })]
    fn test_program_execute(input: &str) {
        let program = parse_term(input).unwrap();
        let instructions = compile_program(program);
        let mut machine = Machine::new();
        machine.set_code(&instructions);
        run_code(&mut machine);

        let output = writeout(&machine.iter_heap().collect::<Vec<_>>());
        assert_display_snapshot!(case(
            input,
            output
        ));
    }
}
