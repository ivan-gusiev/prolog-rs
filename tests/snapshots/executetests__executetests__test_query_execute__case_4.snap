---
source: tests/executetests.rs
expression: "case(input, output)"
---
a
-----
000	<STR,1>
001	a/0

